### ISSUE TITLE
[Eager] Fix for determining input / output shape of the model prior to Model.fit()

### ISSUE THREAD

**Issue by titu1994** [tensorflow/tensorflow#19241]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 Home Edition
- **TensorFlow installed from (source or binary)**: Windows binary
- **TensorFlow version (use command below)**: Github version 'v1.8.0-0-g93bc2e2072' 1.8.0TF (GPU) 
- **Python version**: 3.5
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: CUDA 9.0 / cuDNN 7.0.5
- **GPU model and memory**: NVIDIA GTX 980M
- **Exact command to reproduce**: Provided below as a standalone script

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
The issue is that in Eager mode, the two attributes of the Model subclass, `inputs` and `outputs` is undetermined until the first call to `Model.fit(...)`.

When attempting to determine this prior to fitting the model, in the script at location **tensorflow/python/keras/_impl/keras/engine/training.py**, the entire input dataset `X` (passed to .fit(...)) is provided as the input `x` in Line 684 `if not self.inputs: self._set_inputs(x)` inside `_standardize_user_data`.

Due to this, in eager execution mode, this call is deferred to `_eager_set_inputs(inputs)`. Here, `inputs` is the entire dataset numpy matrix / tensor, and a `Model.call(inputs)` is performed at line 909. 

Since the entire dataset is unable to fit in GPU memory for smaller GPU devices, it causes an OOM error.

However, to determine the input / output shape/s of a model, a single sample tensor is sufficient.

The below fix shows that the solution is adequate, and can be implemented by simply extracting a single sample of the entire dataset to determine the input / output shape/s of a model during eager execution.

**Note : **
In order to provide indicators to identify where the error occurs, I modified the above mentioned script to print 2 logs to the console, to describe the shape of the "inputs" parameter inside `_eager_set_inputs(inputs)`. The lines `Inside training._eager_set_inputs ...` and `***** Providing entire dataset into call *****` are the above logs.

### Source code / logs

```python
import os
import numpy as np

import tensorflow as tf
from tensorflow.python.keras.datasets import mnist
from tensorflow.contrib.eager.python import tfe

# enable eager mode
tf.enable_eager_execution()
tf.set_random_seed(0)
np.random.seed(0)

# constants
batch_size = 128
epochs = 10
num_classes = 10

# dataset loading
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.astype('float32') / 255.
x_test = x_test.astype('float32') / 255.
x_train = x_train.reshape((-1, 28, 28, 1))
x_test = x_test.reshape((-1, 28, 28, 1))

# one hot encode the labels. convert back to numpy as we cannot use a combination of numpy
# and tensors as input to keras
y_train_ohe = tf.one_hot(y_train, depth=num_classes).numpy()
y_test_ohe = tf.one_hot(y_test, depth=num_classes).numpy()

print('x train', x_train.shape)
print('y train', y_train_ohe.shape)
print('x test', x_test.shape)
print('y test', y_test_ohe.shape)

class CNN(tf.keras.Model):

    def __init__(self, num_classes):
        super(CNN, self).__init__()

        self.cnn1 = tf.keras.layers.Conv2D(16, (5, 5), padding='same', strides=(2, 2))
        self.bn1 = tf.keras.layers.BatchNormalization()
        self.cnn2 = tf.keras.layers.Conv2D(32, (5, 5), padding='same', strides=(2, 2))
        self.bn2 = tf.keras.layers.BatchNormalization()
        self.pool = tf.keras.layers.GlobalAveragePooling2D()
        self.classifier = tf.keras.layers.Dense(num_classes)

    def call(self, inputs, training=None, mask=None):
        # Used to print out the input shape of the entire dataset prior to training loop
        print(inputs.shape)

        x = self.cnn1(inputs)
        x = self.bn1(x)
        x = tf.nn.relu(x)  # layer 1
        x = tf.nn.relu(self.bn2(self.cnn2(x))) # layer 2
        x = self.pool(x)
        output = self.classifier(x)

        # softmax op does not exist on the gpu, so always use cpu
        with tf.device('/cpu:0'):
            output = tf.nn.softmax(output)

        return output


device = '/cpu:0' if tfe.num_gpus() == 0 else '/gpu:0'

with tf.device(device):
    # build model and optimizer
    model = CNN(num_classes)
    model.compile(optimizer=tf.train.AdamOptimizer(0.001), loss='categorical_crossentropy',
                  metrics=['accuracy'])
    
    # suggested fix ; can be incorporated inside `_eager_set_inputs` or `_set_input`
    # Fix = Use exactly one sample from the provided input dataset to determine 
    # input/output shape/s for the model

    # dummy_x = np.zeros((1, 28, 28, 1))
    # model._set_inputs(dummy_x)

    # train
    model.fit(x_train, y_train_ohe, batch_size=batch_size, epochs=epochs,
              validation_data=(x_test, y_test_ohe), verbose=2)

    # evaluate on test set
    scores = model.evaluate(x_test, y_test_ohe, batch_size, verbose=2)
    print("Final test loss and accuracy :", scores)

```

Truncated log without the fix : (Only tensor allocation summary dump after OOM is truncated)
```python
2018-05-12 00:22:44.208127: I T:\src\github\tensorflow\tensorflow\core\platform\cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2018-05-12 00:22:45.385242: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1356] Found device 0 with properties: 
name: GeForce GTX 980M major: 5 minor: 2 memoryClockRate(GHz): 1.1265
pciBusID: 0000:01:00.0
totalMemory: 4.00GiB freeMemory: 3.32GiB
2018-05-12 00:22:45.385822: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1435] Adding visible gpu devices: 0
2018-05-12 00:22:45.838116: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:923] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-05-12 00:22:45.838626: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:929]      0 
2018-05-12 00:22:45.838912: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:942] 0:   N 
2018-05-12 00:22:45.839762: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1053] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 3050 MB memory) -> physical GPU (device: 0, name: GeForce GTX 980M, pci bus id: 0000:01:00.0, compute capability: 5.2)
x train (60000, 28, 28, 1)
y train (60000, 10)
x test (10000, 28, 28, 1)
y test (10000, 10)

Inside training._eager_set_inputs ip shape/s [(60000, 28, 28, 1)]  <= Here the input is the entire dataset
***** Providing entire dataset into call *****
(60000, 28, 28, 1)

2018-05-12 00:22:58.125845: W T:\src\github\tensorflow\tensorflow\core\common_runtime\bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 717.77MiB.  Current allocation summary follows.

<<< Truncated >>>

2018-05-12 00:22:58.229153: W T:\src\github\tensorflow\tensorflow\core\framework\op_kernel.cc:1318] OP_REQUIRES failed at fused_batch_norm_op.cc:263 : Resource exhausted: OOM when allocating tensor with shape[60000,16,14,14] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
Traceback (most recent call last):
  File "D:/Users/Yue/PycharmProjects/eager-tutorials/tutorials/04_cnn.py", line 83, in <module>
    validation_data=(x_test, y_test_ohe), verbose=2)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\engine\training.py", line 1147, in fit
    batch_size=batch_size)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\engine\training.py", line 685, in _standardize_user_data
    self._set_inputs(x)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\engine\training.py", line 871, in _set_inputs
    self._eager_set_inputs(inputs)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\engine\training.py", line 914, in _eager_set_inputs
    ops.convert_to_tensor(inputs, dtype=K.floatx()))
  File "D:/Users/Yue/PycharmProjects/eager-tutorials/tutorials/04_cnn.py", line 55, in call
    x = self.bn1(x)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\engine\base_layer.py", line 314, in __call__
    output = super(Layer, self).__call__(inputs, *args, **kwargs)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\layers\base.py", line 717, in __call__
    outputs = self.call(inputs, *args, **kwargs)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\layers\normalization.py", line 113, in call
    output = super(BatchNormalization, self).call(inputs, training=training)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\layers\normalization.py", line 501, in call
    outputs = self._fused_batch_norm(inputs, training=training)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\layers\normalization.py", line 396, in _fused_batch_norm
    training, _fused_batch_norm_training, _fused_batch_norm_inference)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\layers\utils.py", line 206, in smart_cond
    pred, true_fn=true_fn, false_fn=false_fn, name=name)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\framework\smart_cond.py", line 56, in smart_cond
    return false_fn()
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\layers\normalization.py", line 393, in _fused_batch_norm_inference
    data_format=self._data_format)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\ops\nn_impl.py", line 904, in fused_batch_norm
    name=name)
  File "D:\Users\Yue\Anaconda3\lib\site-packages\tensorflow\python\ops\gen_nn_ops.py", line 3804, in _fused_batch_norm
    _six.raise_from(_core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[60000,16,14,14] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc [Op:FusedBatchNorm]

```

Partial log with the aforementioned fix
```python
2018-05-12 00:27:23.607020: I T:\src\github\tensorflow\tensorflow\core\platform\cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2018-05-12 00:27:24.748056: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1356] Found device 0 with properties: 
name: GeForce GTX 980M major: 5 minor: 2 memoryClockRate(GHz): 1.1265
pciBusID: 0000:01:00.0
totalMemory: 4.00GiB freeMemory: 3.32GiB
2018-05-12 00:27:24.748771: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1435] Adding visible gpu devices: 0
2018-05-12 00:27:25.196921: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:923] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-05-12 00:27:25.197231: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:929]      0 
2018-05-12 00:27:25.197570: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:942] 0:   N 
2018-05-12 00:27:25.197993: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1053] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 3050 MB memory) -> physical GPU (device: 0, name: GeForce GTX 980M, pci bus id: 0000:01:00.0, compute capability: 5.2)
x train (60000, 28, 28, 1)
y train (60000, 10)
x test (10000, 28, 28, 1)
y test (10000, 10)

Inside training._eager_set_inputs ip shape/s [(1, 28, 28, 1)]  <= Here the "dataset" is the dummy batch
***** Providing entire dataset into call *****
(1, 28, 28, 1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
(128, 28, 28, 1)
(128, 28, 28, 1)
(128, 28, 28, 1)
(128, 28, 28, 1)
(128, 28, 28, 1)
... repeated for the entire training loop over the batch.
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cy89** [comment]
@alextp can you PTAL?

**Comment by alextp** [comment]
@anj-s I think this is in code you wrote

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 50 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 65 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 80 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 95 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 110 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 125 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @anj-s: It has been 140 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yourtheron** [comment]
I have a very similar, but it is related to saving a custom model as well.
I built a custom layer that was applied in a custom model as follows
```
class C3BR(tf.keras.Model):
    ''' 3D Convolution + Batch Normalisation + Relu '''
    def __init__(self, filterNum, kSize, strSize, padMode):
        super(C3BR, self).__init__()
        self.conv = layers.Conv3D(filters=filterNum, kernel_size=kSize, strides=strSize, padding=padMode, data_format='channels_first')
        self.BN = layers.BatchNormalization(axis=1)
    
    def call(self, inputs, ifTrain):
        x = self.conv(inputs)
        if ifTrain == True:
            x= self.BN(x)
        return activations.relu(x)
        
class SimpleUNet(tf.keras.Model):
    """
    Serialise basic units so as to build up a double-layered encoder-decoder U-Net
    Input:
        inDim: (for initialisation) [modaility/channel, tensor dimensions]
        classNum: background included
        name: name for the net
        inputs: 5D tf tensor of [mbSize, modaility/channel, tensor dimensions]. Inputs must be organised into channel first order
    Returns:
        outputs: 5D tf tensor of [mbSize, classNum, tensor dimensions]
    """
    def __init__(self, inDim, classNum, name='SimpleUNet', **kwarg):
        super(SimpleUNet, self).__init__(name=name, **kwarg)
#    def __init__(self, inDim, classNum, name='SimpleUNet', **kwargs):        
#        super(SimpleUNet, self).__init__(name=name, **kwargs)
        self.inDim = inDim
        self.classNum = classNum
        dimEnSt1End = np.array(inDim)[1:]-2-2
        dimEnSt2Ed = dimEnSt1End/2-2-2
        dimBridgeEnd = (dimEnSt2Ed/2-2-2)*2
        dimDEStd1End = (dimBridgeEnd-2-2)*2
        self.outDim = dimDEStd1End-2-2-2
        temp = ((dimEnSt2Ed - dimBridgeEnd)/2).astype('int32')
        crop3d1 = tuple(np.tile(temp, (2, 1)).T)
        temp = ((dimEnSt1End - dimDEStd1End)/2).astype('int32')
        crop3d2 = tuple(np.tile(temp, (2, 1)).T)

        self.en_st1_cbr1 = C3BR(32, 3, 1, 'valid')
        self.en_st1_cbr2 = C3BR(64, 3, 1, 'valid')
        self.en_st2_mp = layers.MaxPooling3D(pool_size=(2, 2, 2), strides=(2, 2, 2), padding='valid', data_format='channels_first')
        self.en_st2_cbr1 = C3BR(64, 3, 1, 'valid')
        self.en_st2_cbr2 = C3BR(128, 3, 1, 'valid')
        self.bridge_mp = layers.MaxPooling3D(pool_size=(2, 2, 2), strides=(2, 2, 2), padding='valid', data_format='channels_first')
        self.bridge_cbr1 = C3BR(128, 3, 1, 'valid')
        self.bridge_cbr2 = C3BR(256, 3, 1, 'valid')    
        self.bridge_tconv1 = layers.Conv3DTranspose(256, 2, strides=2, padding='valid', data_format='channels_first')
        self.de_3dcrop1 = layers.Cropping3D(crop3d1, data_format='channels_first')
        self.de_st1_cbr1 = C3BR(256, 3, 1, 'valid')
        self.de_st1_cbr2 = C3BR(128, 3, 1, 'valid')    
        self.de_st1_tconv1 = layers.Conv3DTranspose(128, 2, strides=2, padding='valid', data_format='channels_first')
        self.de_3dcrop2 = layers.Cropping3D(crop3d2, data_format='channels_first')
        self.de_st2_cbr1 = C3BR(64, 3, 1, 'valid')
        self.de_st2_cbr2 = C3BR(64, 3, 1, 'valid') 
        self.final_conv3D = layers.Conv3D(filters=self.classNum, kernel_size=3, strides=1, padding='valid', data_format='channels_first')
                
    @tf.function
    def call(self, inputs, ifTrain=True):
        # At present, do not use x = layers.Input(shape=self.inDim), because loads of issues about calling .numpy(), which even
        # exist in tf's own functions
        x = self.en_st1_cbr1(inputs, ifTrain)
        xEnSt1End = self.en_st1_cbr2(x, ifTrain)
        x= self.en_st2_mp(xEnSt1End)
        x= self.en_st2_cbr1(x, ifTrain)
        xEnSt2Ed = self.en_st2_cbr2(x, ifTrain)
        x = self.bridge_mp(xEnSt2Ed)        
        x = self.bridge_cbr1(x, ifTrain)
        x = self.bridge_cbr2(x, ifTrain)      
        xBridgeEnd = self.bridge_tconv1(x)
        xCrop1 = self.de_3dcrop1(xEnSt2Ed)
# printing information, that is where tf raises an error message
        print(xBridgeEnd.shape)
        print(xCrop1.shape)
        x = layers.concatenate([xBridgeEnd, xCrop1], axis=1)
        x = self.de_st1_cbr1(x, ifTrain)
        x = self.de_st1_cbr2(x, ifTrain)
        xDeSt1End = self.de_st1_tconv1(x)
        xCrop2 = self.de_3dcrop2(xEnSt1End)
        x = layers.concatenate([xDeSt1End, xCrop2], axis=1)
        #print(x.shape)
        x = self.de_st2_cbr1(x, ifTrain)
        x = self.de_st2_cbr2(x, ifTrain)
        x = self.final_conv3D(x)
        outputs = activations.softmax(x, axis=1)
        
        return outputs
        
    def compute_output_shape(self):
        # Override this function if you want to use the subclassed model in Kera's fit() method.
        # Otherwise, this method is optional.
        return tf.TensorShape(np.append(self.classNum, self.outDim))    

```

Then  I set it up and tried to use a pair of dummy tensors to set its inputs before saving the model
```
inDim = (4, 64, 64, 64)
classNum = 2
TUNet1 = SimpleUNet(inDim, classNum)
x=tf.random.uniform((16, 4, 64, 64, 64))
y=tf.zeros((16, 2, 22, 22, 22), dtype=uint8)
TUNet1._set_input(x, outputs=y)
````
So far, it went well. but when I further call
```
TUNet1.save(r'...\filename', save_format='tf')
```

The following output is given
```
(None, 256, 18, 18, 18)
(None, 128, 18, 18, 18)
(None, 256, 18, 18, 18)
(None, 128, 18, 18, 18)
(None, 256, None, None, None)
(None, 128, 18, 18, 18)
Traceback (most recent call last):

  File "<ipython-input-88-e56e75f05a97>", line 1, in <module>
    TUNet1.save(r'D:\Work\LiVision\Tech\Data\NN\Study_TF\name\weight', save_format='tf')

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\engine\network.py", line 975, in save
    signatures, options)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\save.py", line 115, in save_model
    signatures, options)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save.py", line 74, in save
    save_lib.save(model, filepath, signatures, options)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\saved_model\save.py", line 870, in save
    checkpoint_graph_view)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\saved_model\signature_serialization.py", line 64, in find_function_to_export
    functions = saveable_view.list_functions(saveable_view.root)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\saved_model\save.py", line 141, in list_functions
    self._serialization_cache)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\engine\base_layer.py", line 2422, in _list_functions_for_serialization
    .list_functions_for_serialization(serialization_cache))

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\base_serialization.py", line 91, in list_functions_for_serialization
    fns = self.functions_to_serialize(serialization_cache)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\layer_serialization.py", line 79, in functions_to_serialize
    serialization_cache).functions_to_serialize)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\layer_serialization.py", line 94, in _get_serialized_attributes
    serialization_cache)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\model_serialization.py", line 53, in _get_serialized_attributes_internal
    serialization_cache))

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\layer_serialization.py", line 103, in _get_serialized_attributes_internal
    functions = save_impl.wrap_layer_functions(self.obj, serialization_cache)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 166, in wrap_layer_functions
    '{}_layer_call_and_return_conditional_losses'.format(layer.name))

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 494, in add_function
    self.add_trace(*self._input_signature)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 413, in add_trace
    trace_with_training(True)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 411, in trace_with_training
    fn.get_concrete_function(*args, **kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 538, in get_concrete_function
    return super(LayerCall, self).get_concrete_function(*args, **kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\eager\def_function.py", line 776, in get_concrete_function
    self._initialize(args, kwargs, add_initializers_to=initializer_map)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\eager\def_function.py", line 408, in _initialize
    *args, **kwds))

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\eager\function.py", line 1848, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\eager\function.py", line 2150, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\eager\function.py", line 2041, in _create_graph_function
    capture_by_value=self._capture_by_value),

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\framework\func_graph.py", line 915, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\eager\def_function.py", line 358, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 515, in wrapper
    ret = method(*args, **kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 475, in wrap_with_training_arg
    return call_fn(*args, **kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\saving\saved_model\save_impl.py", line 557, in call_and_return_conditional_losses
    return layer_call(inputs, *args, **kwargs), layer.get_losses_for(inputs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\eager\function.py", line 2658, in bound_method_wrapper
    return wrapped_fn(*args, **kwargs)

  File "D:\TProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\framework\func_graph.py", line 905, in wrapper
    raise e.ag_error_metadata.to_exception(e)

ValueError: in converted code:
    relative to D:\:

    ork\LiVision\Tech\Product\Alg\Alg-Python\MedImgProc\MedNN.py:87 call  *
        x = layers.concatenate([xBridgeEnd, xCrop1], axis=1)
    ProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\layers\merge.py:705 concatenate
        return Concatenate(axis=axis, **kwargs)(inputs)
    ProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\engine\base_layer.py:817 __call__
        self._maybe_build(inputs)
    ProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\engine\base_layer.py:2141 _maybe_build
        self.build(input_shapes)
    ProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\utils\tf_utils.py:306 wrapper
        output_shape = fn(instance, input_shape)
    ProgramFiles\Anaconda3\envs\keras-gpu\lib\site-packages\tensorflow_core\python\keras\layers\merge.py:391 build
        'Got inputs shapes: %s' % (input_shape))

    ValueError: A `Concatenate` layer requires inputs with matching shapes except for the concat axis. Got inputs shapes: [(None, 256, None, None, None), (None, 128, 18, 18, 18)]

```
I cannot understand why, when saving a model, it calls the model several times, and then suddenly loses the shape in the middle of the stacks of layers.
By the way, I can actually train the model by going through many samples via writing custom training loops using tf.GradientTape() etc. But why such an error occurs when trying to save the model?

**Comment by yourtheron** [comment]
@alextp . Also, related to the issue in [another thread #27120](https://github.com/tensorflow/tensorflow/issues/27120)
I tried to put 
```
xBridgeEnd = layers.Conv3DTranspose(256, 2, strides=2, padding='valid', data_format='channels_first')(x)
```
directly in the SimpleUNet.call() method, tf gives me a
```
ValueError: tf.function-decorated function tried to create variables on non-first call.
```
May I learn the root cause and how to fix it? Thanks

**Comment by yourtheron** [comment]
@alextp It is actually a bug, I think. Please see [my bug report](https://github.com/tensorflow/tensorflow/issues/35441)

**Comment by alextp** [comment]
@yourtheron please file a separate issue for your problem which looks unrelated to this one

**Comment by yourtheron** [comment]
@alextp . Yes, I have issued a bug report.

**Comment by anj-s** [comment]
assigning to @karmel to find the right owner for this issue.

**Comment by mohantym** [comment]
Hi @titu1994!
We are checking to see if you still need help on this issue. We recommend that you upgrade to 2.6 which is latest stable version of TF and let us know if the issue still persists in newer versions. Thanks!

**Comment by titu1994** [comment]
Yes it seems fixed in 2.6. Thanks !

**Comment by mohantym** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)
